### DISCUSSION TITLE
When are bottles built during the release process for a new version of a `homebrew-core` formula?

### DISCUSSION THREAD

**Discussion by rgoldberg** [Homebrew/discussions#5811] [Writing Formulae/Casks] [answered]
### Output of `brew config`

```shell
n/a
```

### Output of `brew doctor`

```shell
n/a
```

### Description of issue

When are bottles built during the release process for a new version of a `homebrew-core` formula?

e.g., [this autogenerated comment on a PR](https://github.com/Homebrew/homebrew-core/pull/200596#issuecomment-2530236165) links to a workflow run which includes [this job](https://github.com/Homebrew/homebrew-core/actions/runs/12249172963/job/34170034089) whose `Pull PR bottles` step obtains bottles from the artifacts of the PR.

I couldn't, however, discern where in prior workflows or in that workflow the bottles were actually built.

Any help would be appreciated.

Thanks.

**Comment by SMillerDev** [comment]
They are the build that happens before. Those bottles are then used to run tests on for CI

**Reply by rgoldberg** [reply]
@SMillerDev Can you let me know the GitHub workflow, job, and step that actually builds the bottles?

I couldn’t find what workflow builds them.

Maybe I completely missed it or maybe that workflow is visible only to Homebrew team members…

I’m trying to understand what options I can use to affect how the bottles are built, so seeing the code that builds them would be helpful.

Thanks again.

**Reply by SMillerDev** [reply]
You can not affect in any way outside of the Formula methods how bottles are built. Doing so would be a severe security issue in the homebrew project.

**Reply by alebcay** [reply, marked as answer]
For transparency - the relevant GitHub Actions step is [here](https://github.com/Homebrew/homebrew-core/blob/4dda18530bdbdd765c9459ea1c9384c4a0925b2d/.github/workflows/tests.yml#L235-L248).

The code for the `brew test-bot` command is at https://github.com/Homebrew/homebrew-test-bot. It acts as a harness around various `brew` commands to test and build the entire lifecycle of a formula. Depending on what you're trying to accomplish, you could consider forking it if you want to build formulae for a custom tap with a modified process.

Within homebrew-core, the aspects of the build process which are intended to be configurable for individual formulae are exposed via the DSL (in formula code).

**Reply by rgoldberg** [reply]
@alebcay Thanks for the links.

For the merged PRs that I’ve looked through, the logs for the `CI` workflow do not include any jobs named `tests`. I see many other jobs from that workflow. It’s possible that `tests` was somehow renamed as `${{ matrix.name }}` in the job list, and not run. (if that was the case, however, how did the later workflow get the bottles from the PR’a artifacts, and how would the checksum for the bottles have changed from their values in the previous version of the formula?)

Can you see a `tests` job in the `CI` workflow [here](https://github.com/Homebrew/homebrew-core/pull/200596/checks)? If so, I guess I don’t have the requisite permissions to see it.

Thanks for the help.

**Reply by alebcay** [reply]
The `tests` job runs as a matrix - you can see this on the flow diagram in the summary view [here](https://github.com/Homebrew/homebrew-core/actions/runs/12395154732), for instance.

The names of the individual jobs in the matrix are the platform it runs on, e.g. "macOS 15-arm64". Note that this matrix job may not always be run (previous steps may cause the `tests` job to be skipped).

The `tests` job runs as part of the pre-merge workflow. If you're looking at the last workflow for a PR which has already been merged, you probably won't see it there, since it is a different workflow that runs upon merge/acceptance.

**Reply by rgoldberg** [reply]
@alebcay Thangs again for the info.

I found the [logs for the workflow, job, and step that built the bottles in which I am interested](https://github.com/Homebrew/homebrew-core/actions/runs/12245688373/job/34160209738?pr=200596).

Earlier, I was using the GitHub iPhone app, which only seemed to let me see the workflows for the most recent commit of in the PR, which was for updating the bottle checksums, not for the submitted PR changes.

On the desktop GitHub site, I can switch commits with a dropdown on the left. I didn't see an equivalent dropdown on the iPhone.

That's what prevented me from finding the logs.
